Rewrite of Reference Transformations (#667)
smart-data-lake/smart-data-lake: docs/reference/transformations.md

:::

#### Example
As a simple example, we add a WhitelistTransformer to a CopyAction.
This will result in only the given columns being selected from the input Data Object and any other columns being discarded.

```
SampleWhitelistAction {
    type = CopyAction
    inputId = someInput
    outputId = someOutput
    transformers = [{
        type = WhitelistTransformer
        columnWhitelist = [includedColumn1,anotherColumn2]
    }]
}
```
Note the square brackets: _transformers_ is an array as you can define multiple transformers.

The parameter `columnWhitelist` is dependent on the type of the transformer.
For details about given parameters / options, please see the [Configuration Schema Viewer](../../JsonSchemaViewer).

## Predefined Transformations
SDLB comes with commonly used transformers out-of-the-box.

The following predefined 1-to-1 transformations are supported:

| Transformer | Description                                                                                           |
| ----------- |-------------------------------------------------------------------------------------------------------|

> | AdditionalColumnsTransformer | Add additional columns to the DataFrame by extracting information from the context                    |

Reviewer: ```suggestion
| AdditionalColumnsTransformer | Add additional columns to the DataFrame by extracting information from the context  or derive new columns from existing columns  |
```

| BlacklistTransformer | Apply a column blacklist to discard columns                                                           |
| DataValidationTransformer | validates DataFrame with a user defined set of rules and creates column with potential error messages |
| DecryptColumnsTransformer | Decrypts specified columns using AES/GCM algorithm                                                    |
| EncryptColumnsTransformer | Encrypts specified columns using AES/GCM algorithm                                                    |
| FilterTransformer | Filter DataFrame with expression                                                                      |

> | SparkRepartitionTransformer | Repartions a DataFrame                                                                                |

Reviewer: ```suggestion
| SparkRepartitionTransformer | Repartitons a DataFrame                                                                                |
```

| StandardizeColNamesTransformer | Used to standardize column names according to configurable rules                                      |

> | StandardizeDatatypesTransformer | Standardize data types of a DataFrame (decimal to corresponding int / float)                          |

Reviewer: ```suggestion
| StandardizeSparkDatatypesTransformer| Standardize data types of a DataFrame (decimal to corresponding int / float)                          |
```

|----------------------------------|-----------------------------------------------------------|
| SQLDfTransformer                 | SQL Transformation 1-to-1                                 |
| SQLDfsTransformer                | SQL Transformation many-to-many                           |
| ScalaClassGenericDfTransformer   | Spark DataFrame transformation in Scala 1-to-1            |
| ScalaClassGenericDfsTransformer  | Spark DataFrame transformation in Scala many-to-many      |
| ScalaClassSnowparkDfTransformer  | Snowpark (Snowflake) transformation in Scala 1-to-1       |
| ScalaClassSnowparkDfsTransformer | Snowpark (Snowflake) transformation in Scala many-to-many |
| ScalaClassSparkDfTransformer     | Spark DataFrame transformation in Scala 1-to-1            |
| ScalaClassSparkDfsTransformer    | Spark DataFrame transformation in Scala many-to-many      |
| ScalaClassSparkDsTransformer     | Spark DataSet transformation in Scala 1-to-1              |
| ScalaClassSparkDsNTo1Transformer | Spark DataFrame transformation in Scala many-to-one       |
| ScalaNotebookSparkDfTransformer  | Loads custom code from a Notebook                         | 

> | PythonCodeSparkDfTransformer | Python implementation (using PySpark)                     |

Reviewer: ```suggestion
| PythonCodeSparkDfTransformer | Spark DataFrame transformation in Python 1-to-1 (using PySpark)  |
| PythonCodeSparkDfsTransformer | Spark DataFrame transformation in Python many-to-many (using PySpark) |
```

There are usually two variants, one for 1-to-1 transformations called _DfTransformer_ and one
for many-to-many transformations called _DfsTransformer_.

:::info
The type of the transformer needs to match your action. 
This is also apparent in the [Configuration Schema Viewer](../../JsonSchemaViewer):  
1-to-1 transformers are listed under 1-to-1 actions, i.e. CopyAction.  
Many-to-many transformers are only listed under many-to-many actions, i.e. CustomDataFrameAction.
:::

### SQL
Spark SQL is probably the easiest way to write a custom transformation, directly in your HOCON configuration.
All input Data Objects are available as tables to select from.
Because of naming limitations, all special characters are replaced by an underscore. 

> So an input Data Object called `Table-With-Hyphen` becomes `table_with_hyphen` inside the SQL query.

Reviewer: ```suggestion
All input Data Objects are available in the select statement with following naming:
- special characters are replaced by an underscore
- a postfix `_sdltemp` is added.
So an input Data Object called `table-with-hyphen` becomes `table_with_hyphen_sdltemp` inside the SQL query.
To simplify this you can also use the special token %{inputViewName} for 1-to-1 transformations, or ${inputViewName_<inputDataObjectName>} for n-to-m transformations, that will be replaced with the correct name at runtime.
```

##### SQL 1-to-1
1-to-1 transformations use type SQLDfTransformer. 

Let's assume we have an input Data Object called dataObject1. 
We can then write a SQL transformation directly in our HOCON configuration:
```
transformers = [{
  type = SQLDfTransformer

>   sqlCode = "select id, count(*) from dataObject1 group by id"

Reviewer: ```suggestion
  sqlCode = "select id, count(*) from %{inputViewName} group by id"
```

}]
```

The SQL code gets executed in Spark SQL so you can use all available functions.

##### SQL many-to-many
Many-to-many transformations use SQLDfsTransformer (note the additional s in Dfs). 
Now that we have multiple output Data Objects, we need to declare which SQL statements belongs to 
which Data Object. 
Therefore, we now have a map of objectIds and corresponding SQL statements:
```
transformers = [{
  type = SQLDfsTransformer
  code = {
    dataObjectOut1 = "select id, cnt from dataObjectIn1 where group = 'test1'",

>     dataObjectOut2 = "select id, cnt from dataObjectIn1 where group = 'test2'"

Reviewer: ```suggestion
    dataObjectOut1 = "select id, cnt from %{inputViewName_dataObjectIn1} where group = 'test1'",
    dataObjectOut2 = "select id, cnt from %{inputViewName_dataObjectIn1} where group = 'test2'"
```

  }
}
```

### Scala
Once transformations get more complex, it's more convenient to implement them in Scala code. 
In custom Scala code, the whole Spark Dataset API is available. 
It's of course also possible to include additional libraries for your code, 
so anything you can do in Spark Scala, you can do with SDLB. 

> ##### As separate file

Reviewer: ```suggestion
##### As Scala class
```

> If you have a Java/Scala project, it usually makes sense to create separate files for your custom Scala code.
> Any files in your classpath are picked up and can be referenced. 

Reviewer: ```suggestion
If you have a Java/Scala project, it usually makes sense to create separate classes for your custom Scala code.
Any classes in your classpath are picked up and can be referenced. 
```

You will see CustomDataFrameAction as a direct action type, similar to CopyAction or HistorizeAction.

ScalaClassSparkDfsTransformer is the type of your **transformer**. 
It needs to correspond to the action type. A 1-to-1 action, expects a 1-to-1 transformer.
:::

##### In Configuration
If you don't work with a full Java/Scala project, it's still possible to define your transformations 
in your HOCON configuration and compile it at runtime.
In this case, use the transformer type ScalaCodeSparkDfTransformer.

To include the Scala code inline:
```
myActionName {
  type = CopyAction
  inputId = stg-tbl1
  outputId = int-tbl1
  transformers = [{
    type = ScalaCodeSparkDfTransformer
    code = """
      import org.apache.spark.sql.{DataFrame, SparkSession}
      import org.apache.spark.sql.functions.explode
      (session: SparkSession, options: Map[String,String], df: DataFrame, dataObjectId: String) => {
        import session.implicits._
        df.select("col1","col2")
      }
    """
  }]
```
Check the method signature in CustomDfTransformer and CustomDfsTransformer according to your requirements.

### Python

> At the moment, only 1-to-1 transformations are supported in Python.

Reviewer: now also n-to-m :-)

Author: Right, just in time for this review :-)
Changed the text in https://github.com/smart-data-lake/smart-data-lake/pull/667/commits/a820df6a0ccafeb0e1eb702cca732eb0f264fac6. 

Not too sure about the syntax. Would that be a valid example?
```
transformers = [{
  type = PythonCodeDfsTransformer 
  code = """
    |dfResult = inputDfs[df1].join(inputDfs[df2], inputDfs[df1].id == inputDfs[df2].fk, "left")\
    | .select(inputDfs[df1].id, inputDfs[df1].desc, inputDfs[df2].desc)\
    |setOutputDf("outputId"->dfResult)
  """
}]
```

Reviewer: ```
transformers = [{
  type = PythonCodeDfsTransformer 
  code = """
    |dfResult = inputDfs[df1].join(inputDfs[df2], inputDfs[df1].id == inputDfs[df2].fk, "left") \
    |  .select(inputDfs[df1].id, inputDfs[df1].desc, inputDfs[df2].desc)
    |setOutputDf({"outputId": dfResult})
  """
}]
```

Author: Thanks a lot. Added the example and description, please re-check this part: https://github.com/smart-data-lake/smart-data-lake/pull/667/commits/62137a5d6c43661d8c2908db6e5e85e86a4016d3

}]
```

PySpark is initialized automatically and the PySpark session is available under the variables
`sc`, `session` or `sqlContext`. 
Some additional variables are also available:
* `inputDf`: Input DataFrame
* `dataObjectId`: Id of input DataObject as string

The output DataFrame needs to be set with `setOutputDf(df)`.

##### Requirements
Running Python transformations needs some additional setup. 

> In general, Python >= 3.4 is required and PySpark package needs to be installed with a version matching your SDL spark version.

Reviewer: ```suggestion
In general, Python >= 3.4 is required and PySpark package needs to be installed with a version matching your SDL spark version. Further environment variable PYTHONPATH needs to be set to your python environment `.../Lib/site-packages` directory, and pyspark command needs to be accessible from the PATH environment variable.
```

### Options / RuntimeOptions
So far these transformations have been quite static: 
Code written can probably only be used for one specific action. 

For custom transformers, you can therefore provide additional options:
* `options`: static options provided in your HOCON configuration
* `runtimeOptions`: extracted at runtime from the context. 

##### In SQL

> If you want to use runtimeOptions in SQL, the syntax is %{key}:

Reviewer: ```suggestion
If you want to use options in SQL, the syntax is %{key}:
```

```
transformers = [{
  type = SQLDfTransformer

>   sqlCode = "select id, cnt, '%{test}' as test, %{run_id} as last_run_id from dataObject1"

Reviewer: ```suggestion
  sqlCode = "select id, cnt, '%{test}' as test, %{last_run_id} as last_run_id from dataObject1"
```

}]
```

PySpark is initialized automatically and the PySpark session is available under the variables
`sc`, `session` or `sqlContext`. 
Some additional variables are also available:
* `inputDf`: Input DataFrame
* `dataObjectId`: Id of input DataObject as string

The output DataFrame needs to be set with `setOutputDf(df)`.

And many-to-many inline example:
```
transformers = [{
  type = PythonCodeDfsTransformer 
  code = """
    |dfResult = inputDfs[df1].join(inputDfs[df2], inputDfs[df1].id == inputDfs[df2].fk, "left") \
    |  .select(inputDfs[df1].id, inputDfs[df1].desc, inputDfs[df2].desc)
    |setOutputDf({"outputId": dfResult})
  """
}]
```

> In this case, `inputDfs` is an array and `setOutputDf` expects a map.

Reviewer: ```suggestion
In this case, `inputDfs` is a dictionary (dict) and `setOutputDf` expects a dictionary.
```

### Options / RuntimeOptions
So far these transformations have been quite static: 
Code written can probably only be used for one specific action. 

For custom transformers, you can therefore provide additional options:
* `options`: static options provided in your HOCON configuration
* `runtimeOptions`: extracted at runtime from the context. 

##### In SQL
If you want to use options in SQL, the syntax is %{key}:
```
transformers = [{
  type = SQLDfTransformer
  sqlCode = "select id, cnt, '%{test}' as test, %{last_run_id} as last_run_id from dataObject1"
  options = {
    test = "test run"
  }
  runtimeOptions = {
    last_run_id = "runId - 1" // runtime options are evaluated as spark SQL expressions against DefaultExpressionData
  }
}]
```

##### In Scala
If you check the signature of the `transform` method again, you can see that you get more than just the DataFrames
you can manipulate.
You also get a Map[String,String] called `options`. 
This Map contains the combined `options` and `runtimeOptions`. 
So in your custom class, you can read all options and runtimeOptions and use them accordingly to parametrize your code.

##### In Python

> Similarly in Python, in addition to the variables `inputDf` and `dataObjectId`, you get a variable called `options`

Reviewer: ```suggestion
Similarly in Python, in addition to the variables `inputDf` and `dataObjectId` (resp. `inputsDfs'), you get a variable called `options`
```
